neqo-http3/src/connection_client.rs

```diff
@@ -2364,4 +2430,676 @@ mod tests {
+
+        let server_hs = server.process(client_hs.dgram(), now());
+        assert!(server_hs.as_dgram_ref().is_some()); // Should produce ServerHello etc...
+        let server_ignored = server.process(client_0rtt.dgram(), now());
+        assert!(server_ignored.as_dgram_ref().is_none());
+
+        // The server shouldn't receive that 0-RTT data.
+        let recvd_stream_evt = |e| matches!(e, ConnectionEvent::NewStream { .. });
+        assert!(!server.events().any(recvd_stream_evt));
+
+        // Client should get a rejection.
+        let client_out = client.process(server_hs.dgram(), now());
+        assert!(client_out.as_dgram_ref().is_some());
+        let recvd_0rtt_reject = |e| e == Http3ClientEvent::ZeroRttRejected;
+        assert!(client.events().any(recvd_0rtt_reject));
+
+        // ...and the client stream should be gone.
+        let res = client.stream_close_send(request_stream_id);
+        assert!(res.is_err());
+        assert_eq!(res.unwrap_err(), Error::InvalidStreamId);
+
+        // Client will send Setting frame and open new qpack streams.
+        let _ = server.process(client_out.dgram(), now());
+        check_control_qpack_streams(&mut server);
+    }
+
+    fn zero_rtt_server_setting(
+        token: Vec<u8>,
+        expect_cap_instruction: bool,
+    ) -> (Http3Client, Connection) {
+        let mut client = default_http3_client();
+        let mut server = default_server();
+        assert_eq!(client.state(), Http3State::Initializing);
+        client
+            .set_resumption_token(now(), &token)
+            .expect("Set resumption token.");
+        assert_eq!(client.state(), Http3State::ZeroRtt);
+        let out = client.process(None, now());
+
+        assert_eq!(client.state(), Http3State::ZeroRtt);
+        assert_eq!(*server.state(), State::WaitInitial);
+        let out = server.process(out.dgram(), now());
+
+        // Check that control and qpack streams anda SETTINGS frame are received.
+        // Also qpack encoder stream will send "change capacity" instruction because it has
+        // the peer settings already.
+        check_control_qpack_request_streams_resumption(&mut server, expect_cap_instruction, false);
+
+        assert_eq!(*server.state(), State::Handshaking);
+        let out = client.process(out.dgram(), now());
+        assert_eq!(client.state(), Http3State::Connected);
+
+        let _out = server.process(out.dgram(), now());
+        assert_eq!(*server.state(), State::Connected);
+
+        assert!(client.tls_info().unwrap().resumed());
+        assert!(server.tls_info().unwrap().resumed());
+
+        (client, server)
+    }
+
+    #[test]
+    fn zero_rtt_new_server_setting_are_the_same() {
+        // Connect and get a token
+        let (mut client, mut server) = connect();
+        let token = exchange_token(&mut client, &mut server.conn);
+
+        // Reconnect
+        let (mut client, mut server) = zero_rtt_server_setting(token, true);
```

Comment:
Why not also include `connect()` and `exchange_token()` calls in `zero_rtt_server_setting`?

I think that what you really want is a range of functions:

* make_server(settings) -- make a server with specific settings
* default_server() -- make a server with default settings, which just calls `make_server()`
* connect_with(client, server, token) -- connects a pre-configured client and server and 0-RTT token
* connect() -- default connection

Or something like that.